### DISCUSSION TITLE
Facing an Issue with Hello-World, maven Quickstart

### DISCUSSION THREAD

**Discussion by waynemorphic** [eclipse/jkube#2787] [Q&A] [answered]
After following the Readme steps for a quick start of the Hello-World app, I am facing a few issues.

What works:
- The build is successful with `mvn clean install k8s:build k8s:resource k8s:apply
`
- `minikube start` command runs minikube container in my docker desktop and opens five ports

⚠️  I have shared minikube's docker daemon environment with my shell

What does not work:
- `kubectl get pods` shows:
```
NAME                          READY   STATUS             RESTARTS        AGE
helloworld-588b59fd98-69prl   0/1     CrashLoopBackOff   8 (2m30s ago)   18m
helloworld-85595f97f7-2x2pg   0/1     CrashLoopBackOff   8 (2m24s ago)   3h57m
```

- `kubectl describe pod` outputs:
```
Type     Reason          Age                   From     Message
  ----     ------          ----                  ----     -------
  Normal   Pulling         131m (x26 over 4h1m)  kubelet  Pulling image "helloworld-java:1.16.1"
  Normal   BackOff         70m (x521 over 4h1m)  kubelet  Back-off pulling image "helloworld-java:1.16.1"
  Normal   SandboxChanged  58m                   kubelet  Pod sandbox changed, it will be killed and re-created.
  Normal   Pulling         57m (x2 over 57m)     kubelet  Pulling image "helloworld-java:1.16.1"
  Warning  Failed          57m (x2 over 57m)     kubelet  Failed to pull image "helloworld-java:1.16.1": Error response from daemon: pull access denied for helloworld-java, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
  Warning  Failed          57m (x2 over 57m)     kubelet  Error: ErrImagePull
  Normal   BackOff         57m (x2 over 57m)     kubelet  Back-off pulling image "helloworld-java:1.16.1"
  Warning  Failed          57m (x2 over 57m)     kubelet  Error: ImagePullBackOff
  Normal   SandboxChanged  55m                   kubelet  Pod sandbox changed, it will be killed and re-created.
  Normal   Pulling         53m (x4 over 55m)     kubelet  Pulling image "helloworld-java:1.16.1"
  Warning  Failed          53m (x4 over 55m)     kubelet  Failed to pull image "helloworld-java:1.16.1": Error response from daemon: pull access denied for helloworld-java, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
  Warning  Failed          53m (x4 over 55m)     kubelet  Error: ErrImagePull
  Warning  Failed          53m (x6 over 55m)     kubelet  Error: ImagePullBackOff
  Normal   BackOff         50m (x19 over 55m)    kubelet  Back-off pulling image "helloworld-java:1.16.1"
  Normal   SandboxChanged  41m                   kubelet  Pod sandbox changed, it will be killed and re-created.
  Normal   Pulling         39m (x4 over 41m)     kubelet  Pulling image "helloworld-java:1.16.1"
  Warning  Failed          39m (x4 over 41m)     kubelet  Failed to pull image "helloworld-java:1.16.1": Error response from daemon: pull access denied for helloworld-java, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
  Warning  Failed          39m (x4 over 41m)     kubelet  Error: ErrImagePull
  Warning  Failed          39m (x6 over 41m)     kubelet  Error: ImagePullBackOff
  Normal   BackOff         26m (x61 over 41m)    kubelet  Back-off pulling image "helloworld-java:1.16.1"
  Warning  BackOff         16m (x28 over 22m)    kubelet  Back-off restarting failed container hello-world in pod helloworld-85595f97f7-2x2pg_default(42d30810-88a1-4217-8fae-80d63453a5a1)
  Normal   Pulled          69s (x10 over 22m)    kubelet  Container image "helloworld-java:1.16.1" already present on machine
```

The last line above shows container is present but the status is *CrashLoopBackOff* instead of *Completed*.  What could I be missing?

**Comment by manusa** [comment]
> The last line above shows container is present but the status is CrashLoopBackOff instead of Completed. What could I be missing?

That's weird (considering the age). Have you tried restarting (deleting) the Pod after getting that message? is the sequence for the new Pod the same?

> runs minikube container in my docker desktop

This is the unclear part. I'm not sure what your setup is.

> ⚠️ I have shared minikube's docker daemon environment with my shell

The previous question in the scope of this one. How have you shared the Minikube's Docker daemon? is this the recommended approach for your platform?

Also, from the same console, have you tried to list the images once JKube finishes its image build  process (`docker  images`)? is the image listed?.

**Reply by rohanKanojia** [reply]
@waynemorphic : What logs do the pod show when you try to do `kubectl logs pod/helloworld-588b59fd98-69prl` . Since this application is just printing a simple message and then terminating, maybe it's getting restarted again and again by Kubernetes. 

By default JKube creates a `Deployment` whose restartPolicy is set to `Always`.

What happens if you change Kubernetes controller type to a Job or  a CronJob : 

```xml
      <plugin>
        <groupId>org.eclipse.jkube</groupId>
        <artifactId>kubernetes-maven-plugin</artifactId>
        <version>${jkube.version}</version>
        <configuration>
          <enricher>
              <config>
                <jkube-controller>
                    <type>CronJob</type>
                    <schedule>* * * * *</schedule>
                </jkube-controller>
              </config>
          </enricher>
           ...
```

**Reply by manusa** [reply]
> maybe it's getting restarted again and again

This is highly likely

We need to change this example.

**Reply by rohanKanojia** [reply]
@waynemorphic : Thanks for opening the discussion! Shall we convert this into an issue or create a new issue?

**Comment by waynemorphic** [comment]
> @waynemorphic : Thanks for opening the discussion! Shall we convert this into an issue or create a new issue?

>> @rohanKanojia : Yes we can

**Comment by waynemorphic** [comment]
> Also, from the same console, have you tried to list the images once JKube finishes its image build process (`docker images`)? is the image listed?.
>> @manusa : Yes, the image is listed, but the status is CrashLoopBackOff.



> @waynemorphic : What logs do the pod show when you try to do `kubectl logs pod/helloworld-588b59fd98-69prl` . Since this application is just printing a simple message and then terminating, maybe it's getting restarted again and again by Kubernetes.
> >  @rohanKanojia : This prints the message, "hello world" as expected
>> These are the logs after running `mvn k8s:log`:
```
[INFO] k8s:  [NEW] helloworld-588b59fd98-69prl status: Running 
[INFO] k8s:  [NEW] Tailing log of pod: helloworld-588b59fd98-69prl container: hello-world
[INFO] k8s:  [NEW] Press Ctrl-C to stop tailing the log
[INFO] k8s:  [NEW] 
[INFO] k8s: Hello World!
[INFO] k8s:  [NEW] helloworld-85595f97f7-2x2pg status: Running 
[INFO] k8s:  [OLD] helloworld-85595f97f7-2x2pg status: Running 
[INFO] k8s:  [OLD] helloworld-85595f97f7-2x2pg status: Running 
[INFO] k8s:  [OLD] helloworld-85595f97f7-2x2pg status: Running Ready
[INFO] k8s:  [OLD] helloworld-85595f97f7-2x2pg status: Running 
[INFO] k8s:  [OLD] helloworld-85595f97f7-2x2pg status: Running 
```

**Reply by rohanKanojia** [reply]
@waynemorphic : Quickstart is working, but CrashLoopBackOff (means repeated restarts) can be confusing for people getting started with the project.

**Reply by manusa** [reply]
Quickstart needs to be updated

**Reply by rohanKanojia** [reply, marked as answer]
I've created https://github.com/eclipse/jkube/issues/2788 for updating the quickstart
